staging/src/k8s.io/dynamic-resource-allocation/deviceclass/cache/cache.go

```diff
@@ -0,0 +1,175 @@
+	if !ok {
+		utilruntime.HandleError(fmt.Errorf("object is not a DeviceClass %#v", newObj))
+		return
+	}
+	oldDeviceClass, ok := oldObj.(*resourceapi.DeviceClass)
+	if !ok {
+		utilruntime.HandleError(fmt.Errorf("object is not a DeviceClass %#v", oldObj))
+		return
+	}
+	c.updateMapping(deviceClass, oldDeviceClass)
+}
+
+// OnDelete handles deletion of a device class.
+func (c *ExtendedResourceCache) OnDelete(obj interface{}) {
+	deviceClass, ok := obj.(*resourceapi.DeviceClass)
+	if !ok {
+		tombstone, ok := obj.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			utilruntime.HandleError(fmt.Errorf("couldn't get object from tombstone %#v", obj))
+			return
+		}
+		deviceClass, ok = tombstone.Obj.(*resourceapi.DeviceClass)
+		if !ok {
+			utilruntime.HandleError(fmt.Errorf("tombstone contained object that is not a DeviceClass %#v", obj))
+			return
+		}
+	}
+	c.removeMapping(deviceClass)
+}
+
+// updateMapping updates the cache with the device class mapping.
+// It first removes any existing mappings for this device class to handle
+// ExtendedResourceName changes, then adds the new mappings.
+func (c *ExtendedResourceCache) updateMapping(newDeviceClass, oldDeviceClass *resourceapi.DeviceClass) {
+	if c == nil {
+		return
+	}
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	// Remove old mappings first to handle ExtendedResourceName changes
+	if oldDeviceClass != nil && oldDeviceClass.Spec.ExtendedResourceName != nil {
+		delete(c.mapping, v1.ResourceName(*oldDeviceClass.Spec.ExtendedResourceName))
+		delete(c.mapping, v1.ResourceName(resourceapi.ResourceDeviceClassPrefix+oldDeviceClass.Name))
+	}
+
+	// Add new mappings
+	if newDeviceClass.Spec.ExtendedResourceName != nil {
+		c.mapping[v1.ResourceName(*newDeviceClass.Spec.ExtendedResourceName)] = newDeviceClass.Name
+		c.logger.V(5).Info("Updated extended resource cache for explicit mapping",
+			"extendedResource", *newDeviceClass.Spec.ExtendedResourceName,
+			"deviceClass", newDeviceClass.Name)
+	}
+	// Always add the default mapping
+	defaultResourceName := v1.ResourceName(resourceapi.ResourceDeviceClassPrefix + newDeviceClass.Name)
+	c.mapping[defaultResourceName] = newDeviceClass.Name
+	c.logger.V(5).Info("Updated extended resource cache for default mapping",
+		"extendedResource", defaultResourceName,
+		"deviceClass", newDeviceClass.Name)
+}
+
+// removeMapping removes the device class mapping from the cache.
+// It searches for all mappings to the given device class name and removes them,
+// because the ExtendedResourceName in the deviceClass object may be stale.
+func (c *ExtendedResourceCache) removeMapping(deviceClass *resourceapi.DeviceClass) {
+	if c == nil {
+		return
+	}
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	if deviceClass != nil && deviceClass.Spec.ExtendedResourceName != nil {
+		delete(c.mapping, v1.ResourceName(*deviceClass.Spec.ExtendedResourceName))
+		delete(c.mapping, v1.ResourceName(resourceapi.ResourceDeviceClassPrefix+deviceClass.Name))
```

Comment:
the implicit should be deleted outside deviceClass.Spec.ExtendedResourceName != nil, as long as deviceClass != nil